File: ocaml/lib/walk.ml
(* Recursive directory walker yielding detected_file records. *)

let candidate_extensions =
  [
    ".rgssad";
    ".rgss2a";
    ".rgss3a";
    ".png_";
    ".ogg_";
    ".m4a_";
    ".rpgmvp";
    ".rpgmvo";
    ".rpgmvm";
    ".pak";
    ".png";
    ".ogg";
    ".m4a";
    ".webp";
    ".jpg";
  ]

let has_interesting_extension (path : string) : bool =
  List.mem
    (String.lowercase_ascii (Filename.extension path))
    candidate_extensions

let file_size (path : string) : int64 =
  In_channel.with_open_bin path In_channel.length

type node_kind = Directory | Regular | Other

let node_kind (path : string) : node_kind =
  try
    match (Unix.lstat path).Unix.st_kind with
    | Unix.S_DIR -> Directory
    | Unix.S_REG -> Regular
    | _ -> Other
  with _ -> Other

(* Make [path] relative to [root] (root assumed a prefix). *)
let relative_to (root : string) (path : string) : string =
  let root =
    if Filename.check_suffix root Filename.dir_sep then root
    else root ^ Filename.dir_sep
  in
  let rl = String.length root in
  if String.length path >= rl && String.sub path 0 rl = root then
    String.sub path rl (String.length path - rl)
  else path

let rec all_files (dir : string) : string list =
  match Sys.readdir dir with
  | exception _ -> []
  | names ->
      Array.to_list names
      |> List.concat_map (fun name ->
           let p = Filename.concat dir name in
           match node_kind p with
           | Directory -> all_files p
           | Regular -> [ p ]
           | Other -> [])

let walk (root_dir : string) : Types.detected_file list =
  match node_kind root_dir with
  | Directory ->
    all_files root_dir
    |> List.filter has_interesting_extension
    |> List.filter_map (fun path ->
         match Dispatch.classify path with
         | Some fmt ->
             (match try Some (file_size path) with _ -> None with
             | None -> None
             | Some size_bytes ->
                 Some
                   Types.
                     {
                       abs_path = path;
                       rel_path = relative_to root_dir path;
                       size_bytes;
                       format = fmt;
                     })
          | None -> None)
  | Regular | Other -> []

File: fsharp/src/RpgmDecrypt.Core/Walk.fs
namespace RpgmDecrypt.Core

/// Pure-data walker over the user's game directory.
///
/// Yields `DetectedFile` records for every input file whose extension or
/// magic bytes suggest an RPG Maker asset. Classification is delegated
/// to `Dispatch.classify` so the walker itself is stateless.
module Walk =

    open System.IO

    let candidateExtensions =
        [ ".rgssad"; ".rgss2a"; ".rgss3a"
          ".png_";  ".ogg_";  ".m4a_"
          ".rpgmvp"; ".rpgmvo"; ".rpgmvm"
          ".pak"
          ".png" ; ".ogg" ; ".m4a"; ".webp"; ".jpg" ]

    let private hasInterestingExtension (path: string) : bool =
        let ext = Path.GetExtension(path).ToLowerInvariant()
        candidateExtensions |> List.contains ext

    let walk (rootDir: string) : DetectedFile list =

        if not (Directory.Exists rootDir) then
            []
        else
            let acc = ResizeArray<DetectedFile>()
            for path in Directory.EnumerateFiles(rootDir, "*", SearchOption.AllDirectories) do
                if hasInterestingExtension path then
                    let info = FileInfo path
                    let size = info.Length
                    match Dispatch.classify path with
                    | Some fmt ->
                        let rel = Path.GetRelativePath(rootDir, path)
                        let detected : DetectedFile =
                            { AbsPath = path
                              RelPath = rel
                              SizeBytes = size
                              Format = fmt }
                        acc.Add detected
                    | None -> ()
            List.ofSeq acc

    let walkWithProgress (rootDir: string) (onProgress: string -> unit) : DetectedFile list =
        if not (Directory.Exists rootDir) then
            []
        else
            let acc = ResizeArray<DetectedFile>()
            let mutable visited = 0
            for path in Directory.EnumerateFiles(rootDir, "*", SearchOption.AllDirectories) do
                visited <- visited + 1
                onProgress (sprintf "visited %d" visited)
                if hasInterestingExtension path then
                    let info = FileInfo path
                    let size = info.Length
                    match Dispatch.classify path with
                    | Some fmt ->
                        let rel = Path.GetRelativePath(rootDir, path)
                        let detected : DetectedFile =
                            { AbsPath = path
                              RelPath = rel
                              SizeBytes = size
                              Format = fmt }
                        acc.Add detected
                    | None -> ()
            List.ofSeq acc
